### DISCUSSION TITLE
DNS 出站流量似乎并没有被转发到内置 DNS

### DISCUSSION THREAD

**Discussion by qianbinbin** [v2fly/v2ray-core#2470] [Q&A]
我在 Linux 搭建了透明代理（单纯上网，没有用作主路由或旁路由），个人的理解是标记 `dns-out` 的出站查询应该被转发到 V2Ray 内置的 DNS，但似乎并没有。

简略配置如下：

```json
{
  "inbounds": [
    {
      "tag": "all-in",
      "protocol": "dokodemo-door"
    }
  ],
  "outbounds": [
    {
      "tag": "dns-out",
      "protocol": "dns",
      "streamSettings": {
        "sockopt": {
          "mark": 255
        }
      }
    }
  ],
  "dns": {
    "servers": [
      "https://1.1.1.1/dns-query",
      {
        "address": "223.5.5.5",
        "domains": [
          "geosite:cn"
        ]
      }
    ]
  },
  "routing": {
    "rules": [
      {
        "type": "field",
        "port": 53,
        "inboundTag": [
          "all-in"
        ],
        "outboundTag": "dns-out"
      }
    ]
  }
}
```

但查看日志发现，不管是 `curl baidu.com` 还是 `curl google.com`，V2Ray 配置里的 `223.5.5.5` 和 `https://1.1.1.1/dns-query` 都被无视了，走了 114 的 DNS：

```
2023/04/19 14:44:18 192.168.0.xxx:59114 accepted udp:114.114.114.114:53 [dns-out]
```

114 是我主路由的配置，Linux 把它自动拉取在了 `/etc/resolv.conf` 中，所以看起来就是使用了系统默认的 DNS。

谢谢各位大佬！

**Comment by mydogshitgold** [comment]
I believe that you forgot this part:
```json
{
  "tag": "dns-out",
  "proxySettings": {
    "tag": "proxy_outbound_tag"
  }
}
```

replace `proxy_outbound_tag` with yours.

**Reply by mydogshitgold** [reply]
Additionally, setup a fallback DNS server for "dns-out" in case the request isn't A / AAAA, or bad things will happen.
```json
{
  "tag": "dns-out",
  "settings": {
    "address": "1.1.1.1",
    "port": 53
  },
  "proxySettings": {
    "tag": "proxy_outbound_tag"
  }
}
```

**Reply by qianbinbin** [reply]
Thank you for your reply, but why [this article](https://guide.v2fly.org/app/tproxy.html#%E4%B8%BA-v2ray-%E9%85%8D%E7%BD%AE%E9%80%8F%E6%98%8E%E4%BB%A3%E7%90%86%E7%9A%84%E5%85%A5%E7%AB%99%E5%92%8C-dns-%E5%88%86%E6%B5%81) not configured like this?

**Reply by qianbinbin** [reply]
I tried:

```json
  "settings": {
    "address": "1.1.1.1",
    "port": 53
  },
  "proxySettings": {
    "tag": "proxy"
  }
```

But still seems not working.

**Reply by mydogshitgold** [reply]
What is the debug log? Can you post your full configuration?

**Comment by mydogshitgold** [comment]
Sorry for misunderstanding your v2ray log in the first place. Your DNS traffic was probably not actually sent to `114.114.114.114', if you configured the DNS hijacking correctly.

The configuration listed in [the guide](https://guide.v2fly.org/app/tproxy.html#%E4%B8%BA-v2ray-%E9%85%8D%E7%BD%AE%E9%80%8F%E6%98%8E%E4%BB%A3%E7%90%86%E7%9A%84%E5%85%A5%E7%AB%99%E5%92%8C-dns-%E5%88%86%E6%B5%81) should work, but it wouldn't respond to non-A/AAAA queries. So adding the `"proxySettings"` and `"settings"` is still important. I think your configuration is probably working fine and you just misunderstood the rather misleading logs.

If you set the log level to `"debug"` or `"info"` you should see that v2ray is actually sending the DNS traffic to the server configured in the `"dns"` block.

e.g. I run send a DNS lookup for the domain `parrot.live` to server 123.32.123.33:

```bash
dig +short @123.32.123.33 parrot.live
```

The access log from v2ray reads like this:
```
10.0.0.10:50962 accepted udp:123.32.123.33:53 [dns]
```

But the real packets were sent to `1.1.1.1`, just as I configured it.

![image](https://user-images.githubusercontent.com/123266548/233015739-37fecd94-1ff7-4334-a88b-b9fcbb56dadd.png)

**Reply by qianbinbin** [reply]
Many thanks! I was misled by the logs 😊
